pkg/registry/apis/query/query.go

```diff
@@ -192,15 +195,57 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
 	}
 
 	// Add user headers... here or in client.QueryData
+	pluginID := req.PluginId
+	pluginUID := req.UID
+	comparisonReq := datasourceRequest{
+		PluginId: pluginID,
+		UID:      pluginUID,
+		Request: &v0alpha1.QueryDataRequest{
+			TimeRange: req.Request.TimeRange,
+			Queries:   req.Request.Queries,
+			Debug:     req.Request.Debug,
+		},
+		Headers: req.Headers, // TODO: fix this - this currently isn't used
+	}
+
+	var err error
+	ctx, err = b.client.PreprocessRequest(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	client, err := b.client.GetDataSourceClient(ctx, v0alpha1.DataSourceRef{
-		Type: req.PluginId,
-		UID:  req.UID,
+		Type: pluginID,
+		UID:  pluginUID,
 	})
 	if err != nil {
 		return nil, err
 	}
 
+	code, rsp, dur, err := b.peformQuery(ctx, client, req, pluginID, pluginUID)
+	// Create a response object with the error when missing (happens for client errors like 404)
+	if rsp == nil && err != nil {
+		rsp = &backend.QueryDataResponse{Responses: make(backend.Responses)}
+		for _, q := range req.Request.Queries {
+			rsp.Responses[q.RefID] = backend.DataResponse{
+				Status: backend.Status(code),
+				Error:  err,
+			}
+		}
+	}
+
+	if b.peformComparison(req) {
+		b.log.Debug("running passive mode comparison to multi-tenant")
+		go b.queryDataAndCompare(ctx, pluginID, pluginUID, comparisonReq, code, rsp, dur)
```

Comment:
Nit. Would provide a contextual logger as argument here to get the `traceID` included in the log messages.

```go
ctxLogger := b.log.FromContext(ctx)
``` 

Maybe generally a good idea for any method in `queryAPIBuilder` part of a request.

Reply:
Ah good to know, thank you! Updated